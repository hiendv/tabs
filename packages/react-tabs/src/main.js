export { default as Tab } from './Tab.jsx'
export { default as Tabs } from './Tabs.jsx'
export * from '@hiendv/tabs'
