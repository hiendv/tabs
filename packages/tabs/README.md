# @hiendv/tabs
Utilities & themes for Tabs components

## Installation & Usage
### CDN
```html
<script src="https://cdn.jsdelivr.net/npm/@hiendv/tabs@latest/dist/iife.js" crossorigin="anonymous"></script>
```
```js
console.log(tabs)
/*
{
  styleDark: {tabs: "dark-module_tabs__2D2d-", items: "dark-module_items__vLUYO", item: "dark-module_item__2t_72", item--active: "dark-module_item--active__2BwxE", panel: "dark-module_panel__3HSBq"},
  styleDefault: {tabs: "default-module_tabs__3f18F", items: "default-module_items__4cVRy", item: "default-module_item__2w-4Z", item--active: "default-module_item--active__3kx5I", panel: "default-module_panel__1onVq"}
}
*/
```

### NPM
The package is mostly consumed by internal Vue & React components. Everything is re-exported.
```js
// Vue
import { styleDefault, styleDark } from '@hiendv/vue-tabs'
console.log(styleDefault, styleDark)
```
