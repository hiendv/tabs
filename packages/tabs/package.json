{
  "name": "@hiendv/tabs",
  "version": "0.1.2",
  "module": "dist/es.js",
  "main": "dist/es.js",
  "author": "Hien Dao <hien.dv.neo@gmail.com>",
  "license": "MIT",
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "@hiendv/bem-sass": "^0.1.0"
  },
  "files": [
    "dist",
    "styles",
    "README.md"
  ]
}
