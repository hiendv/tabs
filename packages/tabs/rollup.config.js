import path from 'path'
import postcss from 'rollup-plugin-postcss'

const reslv = p => {
  return path.resolve(__dirname, p)
}

export default [
  {
    input: reslv('src/index.js'),
    output: {
      format: 'es',
      file: reslv('dist/es.js')
    },
    external: id => /\.scss$/.test(id)
  },
  {
    input: reslv('src/index.js'),
    output: {
      name: 'tabs',
      format: 'iife',
      file: reslv('dist/iife.js')
    },
    plugins: [postcss()]
  }
]
