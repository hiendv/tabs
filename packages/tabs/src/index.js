export { default as themeDefault } from '../themes/default.module.scss'
export { default as themeDark } from '../themes/dark.module.scss'
export { setHash } from './location'
