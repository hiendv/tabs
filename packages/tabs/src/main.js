export { default as styleDefault } from '../styles/default.module.scss'
export { default as styleDark } from '../styles/dark.module.scss'
