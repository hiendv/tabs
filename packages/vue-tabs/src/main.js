export { default as Tab } from './Tab'
export { default as Tabs } from './Tabs.vue'
export * from '@hiendv/tabs'
