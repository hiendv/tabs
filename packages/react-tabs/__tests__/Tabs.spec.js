import { setHash } from '@hiendv/tabs'
import { HashRouter as Router, Link } from 'react-router-dom'
import { act, Simulate } from 'react-dom/test-utils'
import React from 'react'
import ReactDOM from 'react-dom'
import { config } from 'react-transition-group'
import { Tabs, Tab } from '../src/main.js'

config.disabled = true
let container

beforeAll(() => {
  container = document.createElement('div')
  document.body.appendChild(container)
})

afterAll(() => {
  document.body.removeChild(container)
  container = null
})

describe('Tabs', () => {
  it('renders', () => {
    act(() => {
      ReactDOM.render(<Tabs>
        <Tab title='First'>1</Tab>
        <Tab title='Second'>2</Tab>
      </Tabs>, container)
    })

    expect(container.innerHTML).toMatchSnapshot()

    const node = container.querySelectorAll('.item')[1]
    Simulate.click(node)
    expect(container.innerHTML).toMatchSnapshot()
  })

  it('renders without contents', () => {
    act(() => {
      ReactDOM.render(<Tabs />, container)
    })
    expect(container.innerHTML).toMatchSnapshot()
  })

  it('renders with custom theme', () => {
    act(() => {
      ReactDOM.render(<Tabs theme={{
        tabs: 'one',
        items: 'two',
        item: 'three',
        'item--active': 'three--active',
        panel: 'four'
      }}>
        <Tab title='First'>1</Tab>
        <Tab title='Second'>2</Tab>
      </Tabs>, container)
    })

    expect(container.innerHTML).toMatchSnapshot()
  })

  it('renders with panel custom bindings', () => {
    act(() => {
      ReactDOM.render(<Tabs show={1}>
        <Tab title='Foo' style={{ backgroundColor: '#f5f5f5' }} className="foobar"><div>One Content</div></Tab>
        <Tab title='Bar'><div>Two Content</div></Tab>
      </Tabs>, container)
    })

    expect(container.innerHTML).toMatchSnapshot()
    Simulate.click(container.querySelectorAll('a')[0])
    expect(container.innerHTML).toMatchSnapshot()
  })

  it('renders with custom navigation', () => {
    act(() => {
      ReactDOM.render(<Tabs show={1} navRenderer={function (items, self) {
        const theme = self.props.theme
        return (
          <nav className={theme.items}>
            {
              items.map(({ className, end, hash, icon, title }, i) => (
                <a
                  className={`${theme.item} ${self.isActive(i) ? theme['item--active'] : ''} ${end ? theme['item--end'] : ''}`}
                  key={i}
                  href={hash || '#'}
                  onClick={(e) => self.handleClick(e, i)}>
                  {title}
                </a>
              ))
            }
          </nav>
        )
      }}>
        <Tab title='First'>1</Tab>
        <Tab title='Second'>2</Tab>
      </Tabs>, container)
    })

    expect(container.innerHTML).toMatchSnapshot()
    Simulate.click(container.querySelectorAll('a')[0])
    expect(container.innerHTML).toMatchSnapshot()
  })

  it('renders with hashes', () => {
    setHash('two')

    act(() => {
      ReactDOM.render(<Tabs>
        <Tab title='First' hash='one'>1</Tab>
        <Tab title='Second' hash='two'>2</Tab>
      </Tabs>, container)
    })

    expect(container.innerHTML).toMatchSnapshot()
    Simulate.click(container.querySelectorAll('a')[0])
    expect(window.location.hash).toEqual('#one')
  })

  it('renders with hashes from react-router', () => {
    setHash('')
    act(() => {
      ReactDOM.render(<Router>
        <Tabs show={2} navRenderer={function (items, self) {
          const theme = self.props.theme
          return (
            <nav className={theme.items}>
              {
                items.map(({ className, end, hash, icon, title }, i) => (
                  <Link
                    className={`${theme.item} ${self.isActive(i) ? theme['item--active'] : ''} ${end ? theme['item--end'] : ''}`}
                    key={i}
                    to={hash || '#'}
                    onClick={(e) => self.handleClick(e, i)}>
                    {title}
                  </Link>
                ))
              }
            </nav>
          )
        }}>
          <Tab title='One' hash='one'>1</Tab>
          <Tab title='Two' hash='two'>2</Tab>
          <Tab title='Three'>3</Tab>
          <Tab title='Four'>4</Tab>
        </Tabs>
      </Router>, container)
    })

    expect(window.location.toString()).toEqual('http://localhost/')
    expect(container.innerHTML).toMatchSnapshot()

    Simulate.click(container.querySelectorAll('a')[0])
    expect(window.location.toString()).toEqual('http://localhost/#one')

    Simulate.click(container.querySelectorAll('a')[1])
    expect(window.location.toString()).toEqual('http://localhost/#two')

    Simulate.click(container.querySelectorAll('a')[2])
    expect(window.location.toString()).toEqual('http://localhost/')

    Simulate.click(container.querySelectorAll('a')[3])
    expect(window.location.toString()).toEqual('http://localhost/')
  })

  it('renders <tab> without contents', () => {
    act(() => {
      ReactDOM.render(<Tabs>
        <Tab title='First' />
        <Tab title='Second'>2</Tab>
      </Tabs>, container)
    })

    expect(container.innerHTML).toMatchSnapshot()

    Simulate.click(container.querySelectorAll('.item')[1])
    expect(container.innerHTML).toMatchSnapshot()
  })

  it('renders ghost <tab>', () => {
    act(() => {
      ReactDOM.render(<Tabs>
        <Tab title='First'>1</Tab>
        <Tab title='Second'>2</Tab>
        <Tab title='Ghost' ghost>Ghost</Tab>
      </Tabs>, container)
    })

    expect(container.innerHTML).toMatchSnapshot()
    Simulate.click(container.querySelectorAll('.item')[2])
    expect(container.innerHTML).toMatchSnapshot()
  })

  it('renders end <tab>', () => {
    act(() => {
      ReactDOM.render(<Tabs>
        <Tab title='First'>1</Tab>
        <Tab title='End' end>End</Tab>
      </Tabs>, container)
    })

    expect(container.innerHTML).toMatchSnapshot()
    Simulate.click(container.querySelectorAll('.item')[1])
    expect(container.innerHTML).toMatchSnapshot()
  })
})
