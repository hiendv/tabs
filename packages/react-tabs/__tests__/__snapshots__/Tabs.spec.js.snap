// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs renders <tab> without contents 1`] = `
<div class="tabs">
  <nav class="items"><a class="item item--active" href="#">First</a><a class="item" href="#">Second</a></nav>
  <div class="panel slide-down"></div>
</div>
`;

exports[`Tabs renders <tab> without contents 2`] = `
<div class="tabs">
  <nav class="items"><a class="item" href="#">First</a><a class="item item--active" href="#">Second</a></nav>
  <div class="panel slide-down slide-down-enter-done">2</div>
</div>
`;

exports[`Tabs renders 1`] = `
<div class="tabs">
  <nav class="items"><a class="item item--active" href="#">First</a><a class="item" href="#">Second</a></nav>
  <div class="panel slide-down">1</div>
</div>
`;

exports[`Tabs renders 2`] = `
<div class="tabs">
  <nav class="items"><a class="item" href="#">First</a><a class="item item--active" href="#">Second</a></nav>
  <div class="panel slide-down slide-down-enter-done">2</div>
</div>
`;

exports[`Tabs renders end <tab> 1`] = `
<div class="tabs">
  <nav class="items"><a class="item" href="#">First</a><a class="item item--active item--end" href="#">End</a></nav>
  <div class="panel slide-down slide-down-enter-done">End</div>
</div>
`;

exports[`Tabs renders end <tab> 2`] = `
<div class="tabs">
  <nav class="items"><a class="item" href="#">First</a><a class="item item--active item--end" href="#">End</a></nav>
  <div class="panel slide-down slide-down-enter-done">End</div>
</div>
`;

exports[`Tabs renders ghost <tab> 1`] = `
<div class="tabs">
  <nav class="items"><a class="item" href="#">First</a><a class="item item--active" href="#">Second</a><a class="item" href="#">Ghost</a></nav>
  <div class="panel slide-down slide-down-enter-done">2</div>
</div>
`;

exports[`Tabs renders ghost <tab> 2`] = `
<div class="tabs">
  <nav class="items"><a class="item" href="#">First</a><a class="item item--active" href="#">Second</a><a class="item" href="#">Ghost</a></nav>
  <div class="panel slide-down slide-down-enter-done">2</div>
</div>
`;

exports[`Tabs renders with custom navigation 1`] = `
<div class="tabs">
  <nav class="items"><a class="item item--active " href="#">First</a><a class="item  " href="#">Second</a></nav>
  <div class="panel slide-down" style="">1</div>
</div>
`;

exports[`Tabs renders with custom navigation 2`] = `
<div class="tabs">
  <nav class="items"><a class="item item--active " href="#">First</a><a class="item  " href="#">Second</a></nav>
  <div class="panel slide-down" style="">1</div>
</div>
`;

exports[`Tabs renders with custom theme 1`] = `
<div class="one">
  <nav class="two"><a class="three" href="#">First</a><a class="three three--active" href="#">Second</a></nav>
  <div class="four slide-down">2</div>
</div>
`;

exports[`Tabs renders with hashes 1`] = `
<div class="tabs">
  <nav class="items"><a class="item item--active" href="one">First</a><a class="item" href="two">Second</a></nav>
  <div class="panel slide-down" style="">1</div>
</div>
`;

exports[`Tabs renders with hashes from react-router 1`] = `
<div class="tabs">
  <nav class="items"><a class="item  " href="#/one">One</a><a class="item  " href="#/two">Two</a><a class="item item--active " href="#/">Three</a><a class="item  " href="#/">Four</a></nav>
  <div class="panel slide-down slide-down-enter-done">3</div>
</div>
`;

exports[`Tabs renders with panel custom bindings 1`] = `
<div class="tabs">
  <nav class="items"><a class="item" href="#">Foo</a><a class="item item--active" href="#">Bar</a></nav>
  <div class="panel slide-down">
    <div>Two Content</div>
  </div>
</div>
`;

exports[`Tabs renders with panel custom bindings 2`] = `
<div class="tabs">
  <nav class="items"><a class="item item--active" href="#">Foo</a><a class="item" href="#">Bar</a></nav>
  <div class="panel slide-down foobar slide-down-enter-done" style="background-color: rgb(245, 245, 245);">
    <div>One Content</div>
  </div>
</div>
`;

exports[`Tabs renders without contents 1`] = `
<div class="tabs">
  <nav class="items"></nav>
  <div class="panel slide-down slide-down-enter-done"></div>
</div>
`;
